src/ImageCarouselReact/widget/ImageCarouselReact.ts

```diff
@@ -0,0 +1,271 @@
+            // That will also show an error in the config check.
+            const guid = this.contextObject ? this.contextObject.getGuid() : "";
+            const constraint = this.entityConstraint.replace("[%CurrentObject%]", guid);
+            const xpathString = "//" + this.imageEntity + constraint;
+            mx.data.get({
+                callback: this.setDataFromObjects.bind(this, callback), // is this callback right?
+                error: (error) => logger.error(this.id + ": An error occurred while retrieving items: " + error),
+                xpath : xpathString,
+            });
+        } else {
+            logger.debug(this.id  + ".getDataFromXpath empty context");
+            this.setDataFromObjects(callback, []);
+        }
+    }
+    private fetchDataFromMicroflow(callback: Function): void {
+        logger.debug(this.id  + ".fetchDataFromMicroflow");
+        if (this.requiresContext && !this.contextObject) {
+            // case there is not context ID the xpath will fail, so it should always show no images.
+            logger.debug(this.id  + ".fetchDataFromMicroflow, empty context");
+            this.setDataFromObjects(callback, []);
+        } else {
+            mx.data.action({
+                callback: this.setDataFromObjects.bind(this, callback),
+                error: (error) => logger.error(this.id  + ": An error occurred while executing microflow: " + error),
+                params: {actionname: this.dataSourceMicroflow,
+                         applyto: this.requiresContext ? "selection" : "none",
+                         guids: this.requiresContext ? [this.contextObject.getGuid()] : [],
+                        },
+            });
+        }
+    }
+    /**
+     * transforms mendix object into item properties and set new state
+     */
+    private setDataFromObjects(callback: Function, objects: mendix.lib.MxObject[]): void {
+        logger.debug(this.id + ".getCarouselItemsFromObject");
+        this.data = objects.map((itemObj): Data => {
+            return {
+                caption: this.captionAttr ? itemObj.get(this.captionAttr) as string : "",
+                description: this.descriptionAttr ? itemObj.get(this.descriptionAttr) as string : "",
+                onClick: {
+                    clickMicroflow: this.callMicroflow,
+                    contextGuid: itemObj.getGuid(),
+                    onClickEvent: this.onClickEvent,
+                    page: this.pageForm,
+                    pageLocation: this.pageLocation,
+                },
+                url: this.getFileUrl(itemObj.getGuid()),
+            };
+        });
+        callback();
+    }
+    /**
+     * Returns a image url from the object Id.
+     */
+    private getFileUrl(objectId: string): string {
+        logger.debug(this.id + ".getFileUrl");
+        let url: string;
+        if (objectId !== null || typeof(objectId) !== "undefined") {
```

Comment:
Object id is required, so if statement could be removed, and string returned without `let` declaration.